redis/src/main/scala/zio/redis/api/PubSub.scala

```diff
@@ -0,0 +1,158 @@
+package zio.redis.api
+
+import zio.redis.Input._
+import zio.redis.Output._
+import zio.redis.ResultBuilder.ResultStreamBuilder
+import zio.redis._
+import zio.schema.Schema
+import zio.stream._
+import zio.{Chunk, IO, Promise, ZIO}
+
+trait PubSub extends RedisEnvironment {
+  import PubSub._
+
+  final def subscribe(channel: String): ResultStreamBuilder[Id] =
+    subscribeWithCallback(channel)(emptyCallback)
+
+  final def subscribe(channel: String, channels: List[String]): ResultStreamBuilder[List] =
+    subscribeWithCallback(channel, channels)(emptyCallback)
+
+  final def subscribeWithCallback(
+    channel: String
+  )(onSubscribe: PubSubCallback): ResultStreamBuilder[Id] =
+    new ResultStreamBuilder[Id] {
+      def returning[R: Schema]: IO[RedisError, Id[Stream[RedisError, R]]] =
+        getSubscribeStreams(channel, List.empty)(onSubscribe).flatMap(extractOne(channel, _))
+    }
+
+  final def subscribeWithCallback(
+    channel: String,
+    channels: List[String]
+  )(onSubscribe: PubSubCallback): ResultStreamBuilder[List] =
+    new ResultStreamBuilder[List] {
+      def returning[R: Schema]: IO[RedisError, List[Stream[RedisError, R]]] =
+        getSubscribeStreams(channel, channels)(onSubscribe)
+    }
+
+  final def unsubscribe(channel: String): IO[RedisError, Promise[RedisError, (String, Long)]] =
+    unsubscribe(List(channel)).flatMap(extractOne(channel, _))
+
+  final def unsubscribe(channels: List[String]): IO[RedisError, List[Promise[RedisError, (String, Long)]]] =
+    RedisPubSubCommand(PubSubCommand.Unsubscribe(channels), codec, pubSub).run
+      .flatMap(
+        ZIO.foreach(_)(stream =>
+          for {
+            promise <- Promise.make[RedisError, (String, Long)]
+            _ <- stream.mapZIO {
+                   case PushProtocol.Unsubscribe(channel, numOfSubscription) =>
+                     promise.succeed((channel, numOfSubscription))
+                   case _ => promise.fail(RedisError.WrongType(s"Cannot handle message except Unsubscribe"))
+                 }.runDrain.fork
+          } yield promise
+        )
+      )
+
+  private def getSubscribeStreams[R: Schema](
+    channel: String,
+    channels: List[String]
+  )(onSubscribe: PubSubCallback) =
+    RedisPubSubCommand(PubSubCommand.Subscribe(channel, channels, onSubscribe), codec, pubSub).run
+      .map(_.map(_.mapZIO {
+        case PushProtocol.Message(_, msg) =>
+          ZIO
+            .attempt(ArbitraryOutput[R]().unsafeDecode(msg)(codec))
+            .refineToOrDie[RedisError]
+            .asSome
```

Comment:
I think that this can be part of the `run` implementation in the `RedisPubSubCommand`.

Reply:
I made `RedisPubSubCommand` takes responsibilities of transforming message types and invoking callback